app/Models/Concerns/HasRoles.php

```diff
@@ -0,0 +1,277 @@
+<?php
+
+namespace App\Models\Concerns;
+
+use App\Events\User\RolesChanged;
+use App\Models\Role;
+use App\User;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Database\Eloquent\ModelNotFoundException;
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
+use Illuminate\Support\Collection;
+
+trait HasRoles
+{
+    use HasPermissions;
+
+    public static function bootHasRoles()
+    {
+        static::deleting(function ($model) {
+            $model->roles()->detach();
+        });
+    }
+
+    /**
+     * A model may have multiple roles.
+     */
+    public function roles(): BelongsToMany
+    {
+        return $this->belongsToMany(Role::class, 'user_roles');
+    }
+
+    /**
+     * Scope the model query to certain roles only.
+     *
+     * @param Builder $query
+     * @param string|array|Collection $roles
+     * @return Builder
+     */
+    public function scopeRole(Builder $query, $roles): Builder
+    {
+        if ($roles instanceof Collection) {
+            $roles = $roles->all();
+        }
+        if (! is_array($roles)) {
+            $roles = [$roles];
+        }
+        $roles = array_map(function ($role) {
+            if ($role instanceof Role) {
+                return $role;
+            }
+            $method = is_numeric($role) ? 'find' : 'findByName';
+
+            return Role::{$method}($role);
+        }, $roles);
+
+        return $query->whereHas('roles', function ($query) use ($roles) {
+            $query->where(function ($query) use ($roles) {
+                foreach ($roles as $role) {
+                    $query->orWhere('roles.id', $role->id);
+                }
+            });
+        });
+    }
+
+    /**
+     * Assign the given role to the model.
+     *
+     * @param Collection|array|string ...$roles
+     *
+     * @return $this
+     */
+    public function assignRole(...$roles): self
+    {
+        $roles = $this->roleInputToIds($roles)->all();
+
+        $changes = $this->roles()->sync($roles, false);
+
+        if ($this instanceof User && collect($changes)->sum(function ($value) {
```

Comment:
I'm not sure what is going on here

Reply:
It is checking that the Model have roles synced is a user model, and it is checking that changes have actually been made in the database. `sync()` returns an array with the keys `created`, `updated` and `deleted` or something like that, with the values of each key being equal to the number of rows mutated.

Reply:
Worth documenting it since I couldn't see it at first sight